internal/internal_workflow.go

```diff
@@ -1611,6 +1611,15 @@ func SetCurrentDetails(ctx Context, details string) {
 	getWorkflowEnvOptions(ctx).currentDetails = details
 }
 
+// SetVersioningBehavior sets the strategy to upgrade this workflow when the default Build ID
```

Comment:
Did we intentionally omit a getter here?

Reply:
Yes, versioning behavior  can be dependent on worker options so if app code branches based on it they will get an NDE. And in general, we wanted to be able to change annotations without needing patching, so no-getter helps there...

Reply:
Is the versioning behaviour written into history?

Reply:
I think you need to be able to get something you may have set. I think this should either be in workflow register options and/or this should be called `SetVersioningBehaviorOverride` (and `GetVersioningBehaviorOverride` will just return what you have called with the "set", not the resolved value).

Reply:
@Quinn-With-Two-Ns My understanding is that it is not part of the history, at least I do nothing special during replay, @ShahabT can you confirm?

Reply:
@cretz we could add the Override suffix, but the emphasis was that most customers will annotate all workflows, and do not set defaults in Workers, since that is safer.
 The Get override, could be useful in the future when we enable changing workflow properties externally (like a reset), maybe we could add it then.
I think the problem of adding to workflow register is that I don't have the context of the running workflow to latch on, so I cannot change the behavior based on arguments, or for dynamic workflows. Also, it may be more clear to have this "annotation" the first line of the workflow definition, together with the code... 
@drewhoskins-temporal any preferences?

Reply:
Looking at the proto changes that @ShahabT did, the versioning behavior has been added to WorkflowTaskCompletedEventAttributes in history. I don't do anything with it (maybe I should?) but that would give the server a placeholder to record the versioning behavior that I send after each task...

Reply:
I do wonder whether the getter should take that into account. I'll defer to y'all on desired behavior of the getter, but I do believe we should not have a setter without a getter (even if we have to add caveats to the behavior of the getter).

Reply:
The server stores the current build id and versioning behavior in Mutable State and in visibility. The previously-used build ids are also stored in visibility -- I'm not sure if previously-used build ids are also stores in Mutable State.

"getting" the current versioning behavior of a given workflow execution could be done by reading the mutable state or visibility entry.

Querying a bunch of workflow executions for a specific versioning behavior and/or build id combination would be a visibility query (we will use this for reachability).

It seems to me that to support a workflow-execution-level versioning behavior override, the server would need to store not just the (versioning behavior, build id) tuple, but also an `is_override` (or some other name) bit, that basically means, `if is_override, ignore whatever the sdk says about my workflow type and do task routing based on what is currently written in my mutable state as an override`. To me it seems like it could be handled on the server side, since that's where the task routing decision is made.

Reply:
> I think of versioning behavior as primarily a workflow type level thing, with a workflow worker level default in case it's not specified at the type level (but from a product perspective, we want to encourage explicitly setting it for each workflow type

If this is the case, it also needs to be added to `workflow.RegisterOptions`.

> "getting" the current versioning behavior of a given workflow execution could be done by reading the mutable state or visibility entry.

We don't want to get from server what's not already in history. We just need to make sure any `SetWhatever` _inside_ a workflow has a corresponding `GetWhatever`. If this just returns the value that was "set" before in the same workflow code and not the actual server value, that's fine (though it does seem like task complete may have some form of the value). The docs of the getter just need to clarify.